### PR TITLE
Parsing of ml weights

### PR DESCRIPTION
**Pull request by schuyann**
Implementing the possibility to save and load the ML-model to and from a file.

### CODE REVIEW

File: src/koopmans/io/_kwf.py [hunk 1]
```diff
@@ -30,7 +36,18 @@ def default(self, obj) -> dict:
...
+        elif (isinstance(obj, Ridge) or isinstance(obj, StandardScaler)):
+            model_params: List[str] = []
+            if isinstance(obj, Ridge):
+                model_params = ['coef_', 'intercept_']
+            elif isinstance(obj, StandardScaler):
+                model_params = ['mean_', 'scale_', 'n_samples_seen_', 'var_']
+            d = {}
+            d['init_params'] = obj.get_params()
+            d['model_params'] = {}
+            for p in model_params:
```
Comment (elinscott) [inline src/koopmans/io/_kwf.py:48]: This would be slightly nicer written as a comprehension

File: src/koopmans/io/_kwf.py [hunk 2]
```diff
@@ -50,6 +67,22 @@ def object_hook(dct):
...
+            if 'scaler' in dct:
+                scaler = StandardScaler()
+                model_dct = dct.pop('scaler')
+                scaler = load_ml_model(scaler, model_dct)
+                dct['scaler'] = scaler
+            if 'model' in dct:
+                model = Ridge()
+                model_dct = dct.pop('model')
+                model = load_ml_model(model, model_dct)
+                dct['model'] = model
```
Comment (elinscott) [inline src/koopmans/io/_kwf.py:85]: Unless there is a good reason not to, let's not place this inside of the `else` statement but instead use the pattern above for `Path` i.e. initially create the dct inside of `KoopmansEncoder` with a `__standardscaler__` or `__ridge__` entry, and then do something similar to line 64

File: src/koopmans/ml/_ml_models.py [hunk 3]
```diff
@@ -18,12 +21,32 @@ def predict(self, x_test: np.ndarray) -> np.ndarray:
...
+        dct = dict(self.__dict__)
+        return dct
+
+    @classmethod
+    def fromdict(cls, dct):
+        return cls(**dct)
+
+    def save_to_file(self, save_dir: Path) -> None:
+        from koopmans.io import write_kwf
+        with open(save_dir / f'model.kwf', 'w') as fd:
```
Comment (elinscott) [inline src/koopmans/ml/_ml_models.py:34]: Do we want to give this a more specific name? How about `self.__class__.__name__.lower()`?

File: src/koopmans/ml/_ml_models.py [hunk 4]
```diff
@@ -18,12 +21,32 @@ def predict(self, x_test: np.ndarray) -> np.ndarray:
...
+            write_kwf(self, fd)
+
+    @classmethod
+    def load_from_file(cls, save_dir: Path):
+        from koopmans.io import read
+        dct = read(save_dir / f'model.kwf')
```
Comment (elinscott) [inline src/koopmans/ml/_ml_models.py:40]: Might need to change this to be consistent with the above suggestion

File: src/koopmans/ml/_ml_models.py [hunk 5]
```diff
@@ -67,51 +100,55 @@ def predict(self, x_test: np.ndarray) -> np.ndarray:
...
+        cls: Type[AbstractPredictor] = predictor_classes[self.type_of_ml_model]
+        if save_dir is None:
+            predictor = cls()
         else:
-            raise ValueError(f"{self.type_of_ml_model} is not implemented as a valid ML model.")
+            if sub_dir is None:
+                predictor = cls.load_from_file(save_dir)
+            else:
+                predictor = cls.load_from_file(save_dir / sub_dir)
+        return predictor
```
Comment (elinscott) [inline src/koopmans/ml/_ml_models.py:122]: Another instance where we don't need a separate argument for `sub_dir`

File: src/koopmans/ml/_ml_models.py [hunk 6]
```diff
@@ -67,51 +100,55 @@ def predict(self, x_test: np.ndarray) -> np.ndarray:
...
                    'no training data has been added so far)'
 
     def todict(self):
-        dct = copy.deepcopy(dict(self.__dict__))
-        items_to_pop = ['model']
-        for item in items_to_pop:
-            dct.pop(item)
+        # Shallow copy
+        dct = dict(self.__dict__)
         return dct
```
Comment (elinscott) [inline src/koopmans/ml/_ml_models.py:136]: Do we need to add a `predictor` entry here to be consistent with `fromdict`?

**Review by elinscott — CHANGES_REQUESTED** [review]
Looking good, just a few minor tweaks needed

File: src/koopmans/ml/_ml_models.py (hunk 6 above)
Comment (schuyann) [reply src/koopmans/ml/_ml_models.py:136]: Always when an instance of this class is initialized with `__init__` the instance also contains a `predictor` entry. Do we still need to explicitly add a `predictor` entry?

Comment (schuyann) [reply src/koopmans/ml/_ml_models.py:137]: I have added `predictor` as an input variable for `__init__` then we don't have to pop it in `fromdict`

**Review by elinscott — APPROVED** [review]
Looks great! Good work. Go ahead and merge when you're ready.

**Pull request merged**